turbo/app/import.go

```diff
@@ -0,0 +1,224 @@
+
+	// Open the file handle and potentially unwrap the gzip stream
+	fh, err := os.Open(fn)
+	if err != nil {
+		return err
+	}
+	defer fh.Close()
+
+	var reader io.Reader = fh
+	if strings.HasSuffix(fn, ".gz") {
+		if reader, err = gzip.NewReader(reader); err != nil {
+			return err
+		}
+	}
+	stream := rlp.NewStream(reader, 0)
+
+	// Run actual the import.
+	blocks := make(types.Blocks, importBatchSize)
+	n := 0
+	for batch := 0; ; batch++ {
+		// Load a batch of RLP blocks.
+		if checkInterrupt() {
+			return fmt.Errorf("interrupted")
+		}
+		i := 0
+		for ; i < importBatchSize; i++ {
+			var b types.Block
+			if err := stream.Decode(&b); err == io.EOF {
+				break
+			} else if err != nil {
+				return fmt.Errorf("at block %d: %v", n, err)
+			}
+			// don't import first block
+			if b.NumberU64() == 0 {
+				i--
+				continue
+			}
+			blocks[i] = &b
+			n++
+		}
+		if i == 0 {
+			break
+		}
+		// Import the batch.
+		if checkInterrupt() {
+			return fmt.Errorf("interrupted")
+		}
+
+		missing := missingBlocks(chainDB, blocks[:i])
+		if len(missing) == 0 {
+			log.Info("Skipping batch as all blocks present", "batch", batch, "first", blocks[0].Hash(), "last", blocks[i-1].Hash())
+			continue
+		}
+
+		// RLP decoding worked, try to insert into chain:
+		missingChain := &core.ChainPack{
+			Blocks:   missing,
+			TopBlock: missing[len(missing)-1],
+			Length:   len(missing),
+		}
+
+		if err := InsertChain(ethereum, missingChain); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func ChainHasBlock(chainDB kv.RwDB, block *types.Block) bool {
+	var chainHasBlock bool
+
+	chainDB.View(context.Background(), func(tx kv.Tx) (err error) {
+		chainHasBlock = rawdb.HasBlock(tx, block.Hash(), block.NumberU64())
+		return nil
+	})
+
+	return chainHasBlock
+}
+
+func missingBlocks(chainDB kv.RwDB, blocks []*types.Block) []*types.Block {
+	var headBlock *types.Block
+	chainDB.View(context.Background(), func(tx kv.Tx) (err error) {
+		hash := rawdb.ReadHeadHeaderHash(tx)
+		number := rawdb.ReadHeaderNumber(tx, hash)
+		headBlock = rawdb.ReadBlock(tx, hash, *number)
+		return nil
+	})
+
+	for i, block := range blocks {
+		// If we're behind the chain head, only check block, state is available at head
+		if headBlock.NumberU64() > block.NumberU64() {
+			if !ChainHasBlock(chainDB, block) {
+				return blocks[i:]
+			}
+			continue
+		}
+
+		if !ChainHasBlock(chainDB, block) {
+			return blocks[i:]
+		}
+	}
+
+	return nil
+}
+
+func InsertChain(ethereum *eth.Ethereum, chain *core.ChainPack) error {
+	sentryControlServer := ethereum.SentryControlServer()
+	// TODO: Not sure if the initialCycle should always be set to false
+	initialCycle := false
+	highestSeenHeader := chain.TopBlock.NumberU64()
+
+	for _, b := range chain.Blocks {
+		sentryControlServer.Hd.AddMinedHeader(b.Header())
+		sentryControlServer.Bd.AddMinedBlock(b)
+	}
+
+	for _, link := range sentryControlServer.Hd.InsertList() {
+		sentryControlServer.Hd.MarkPreverifiedForward(link)
+	}
```

Comment:
I have feeling that - .MarkPreverifiedForward must be called from inside .AddMinedHeader - because mined blocks are "verified enough", and likely we just forgot to do it. 

Is it easy to do it from inside .AddMinedHeader? (if not, fine).